File: 1/2020.10.16/50.Pow(x, n)/java/solution.java
class Solution {

    Map<Integer, Double> memo = new HashMap<>();

    public double myPow(double x, int n) {
        return n > 0 ? pow(x, n) : 1 / pow(x, -n);
    }

    public double pow(double x, int n) {
        if (n == 0) {
            return 1.0d;
        }
        if (memo.containsKey(n)) {
            return memo.get(n);
        }
        double res = pow(x, n / 2);
        res = res * res * (n % 2 == 0 ? 1 : x);
        memo.put(n, res);
        return res;
    }
}

File: 1/2020.10.16/50.Pow(x, n)/python/solution.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        @functools.lru_cache(n)
        def pow(x, n):
            if n == 0:
                return 1
            res = pow(x, n // 2)
            if n % 2 == 0:
                return res * res
            else:
                return res * res * x
        if n < 0:
            return pow(1/x, -n)
        return pow(x, n)
